### PR TITLE
feat: add new authentication flow

### PR DESCRIPTION
**Pull request by matteoscurati**
## **Description**

This PR relates to the integration of new Authentication and User Storage controllers within the extension.

Specifically, the PR implements:
- Logic for integrating Redux with the created controllers
- Necessary hooks to manage actions and their associated loading and error states
- Integration with the onboarding process
- Integration with settings pages where necessary

At the end of the onboarding process, the user will be able to:
- Choose to use MetaMetrics
- Use the Profile Syncing feature (enabling UserStorage) or, through the advanced settings page, choose not to use this feature

If the user chooses to use MetaMetrics or enable Profile Syncing, the user must be authenticated.

Only if the user chooses not to use MetaMetrics and not to enable Profile Syncing can they remain unauthenticated.

Currently, the authentication process continues to use the locally created MetaMetrics ID.

The PR also includes components that allow the user, if they wish, to disable MetaMetrics or Profile Syncing at any time.

[Open in GitHub Codespaces](https://codespaces.new/MetaMask/metamask-extension/pull/24028?quickstart=1)

## **Related issues**

Fixes:

## **Manual testing steps**

1. Follow the onboarding process
2. Check the settings page to turn on/off metametrics and profile syncing

### **After**

This is the new onboarding flow:

https://www.figma.com/file/c7GgNw2kScGrVyRGAPhwEd/Notifications?type=design&node-id=1498-68094&mode=design&t=sFvZbmQXNQaBxs6M-4

## **Pre-merge author checklist**

- [x] I’ve followed [MetaMask Coding Standards](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/CODING_GUIDELINES.md).
- [x] I've completed the PR template to the best of my ability
- [x] I’ve included tests if applicable
- [x] I’ve documented my code using [JSDoc](https://jsdoc.app/) format if applicable
- [x] I’ve applied the right labels on the PR (see [labeling guidelines](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/LABELING_GUIDELINES.md)). Not required for external contributors.

## **Pre-merge reviewer checklist**

- [x] I've manually tested the PR (e.g. pull and build branch, run the app, test code being changed).
- [x] I confirm that this PR addresses all acceptance criteria described in the ticket it closes and includes the necessary testing evidence such as recordings and or screenshots.

### CODE REVIEW

File: ui/ducks/app/app.ts [hunk 1]
```diff
@@ -307,12 +307,12 @@ export default function reduceApp(
     case actionConstants.MODAL_CLOSE:
       return {
         ...appState,
-        modal: Object.assign(
-          appState.modal,
-          { open: false },
-          { modalState: { name: null, props: {} } },
-          { previousModalState: appState.modal.modalState },
-        ),
+        modal: {
```
Comment (matteoscurati) [inline ui/ducks/app/app.ts:310]: This change is to prevent Redux from raising a state management error in some cases.

File: ui/pages/routes/routes.container.js [hunk 2]
```diff
@@ -55,6 +55,16 @@ import Routes from './routes.component';
 
 function mapStateToProps(state) {
   const { activeTab, appState } = state;
+
+  console.log('TEST DATA', {
```
Comment (matteoscurati) [inline ui/pages/routes/routes.container.js:59]: For now, I'm using this console.log to debug the authentication flow

File: ui/ducks/app/app.ts (hunk 1 above)
Comment (danjm) [reply ui/ducks/app/app.ts:337]: These seem really similar to me. How is the before and after code different? How would it fix an error?

File: ui/hooks/metamask-notifications/useProfileSyncing.ts [hunk 3]
```diff
@@ -0,0 +1,156 @@
...
+  return { disableProfileSyncing, loading, error };
+}
+
+/**
+ * Custom hook to check if profile syncing is enabled. This hook uses a selector to retrieve
+ * the profile syncing state from the Redux store.
+ *
+ * @returns An object containing the current state of profile syncing.
+ */
+export function useIsProfileSyncingEnabled(): {
```
Comment (danjm) [inline ui/hooks/metamask-notifications/useProfileSyncing.ts:151]: I see this hook defined here, but I also see this line of code below:
```
const isProfileSyncingEnabled = useSelector(selectIsProfileSyncingEnabled);
```

We could address that specific case, but it raises a question for me. I'm noticing that the use of hooks to wrap a single selector is new in our codebase. On first glance, it adds unneccessary code (e.g. why make a `useIsProfileSyncingEnabled` hooks when we can just directly use `selectIsProfileSyncingEnabled`). Could you share your thoughts on the benefits on the approach taken in this PR with these types of hooks?

File: ui/pages/settings/security-tab/security-tab.component.js [hunk 4]
```diff
@@ -1108,7 +1076,7 @@ export default class SecurityTab extends PureComponent {
           {this.context.t('metrics')}
         </span>
         <div className="settings-page__content-padded">
-          {this.renderMetaMetricsOptIn()}
```
Comment (danjm) [inline ui/pages/settings/security-tab/security-tab.component.js:1111]: Was this change needed for this PR? Or is it just an improvement of the codebase?

File: ui/ducks/app/app.ts (hunk 1 above)
Comment (Gudahtt) [reply ui/ducks/app/app.ts:337]: The previous version was mutating `appState.modal`, but we're not supposed to mutate Redux state. It should be updated immutable, creating new objects rather than mutating.

It seems like a legitimate bug fix to me, though maybe better submitted as a separate PR

File: ui/pages/settings/security-tab/security-tab.component.js (hunk 4 above)
Comment (matteoscurati) [reply ui/pages/settings/security-tab/security-tab.component.js:1111]: @danjm  in the hook I call, I execute a logic whereby if Profile Sync is not active and metametrics is not active, I log out the user. Conversely, if the user is logged out and activates metametrics, I log in the user. This function is necessary for the next implementation of notifications (next PR).

File: ui/hooks/metamask-notifications/useProfileSyncing.ts (hunk 3 above)
Comment (matteoscurati) [reply ui/hooks/metamask-notifications/useProfileSyncing.ts:151]: I agree. In fact, for now, it's nonsensical and a selector would be more than sufficient.

File: ui/ducks/app/app.ts (hunk 1 above)
Comment (matteoscurati) [reply ui/ducks/app/app.ts:337]: Exactly, but I agree it would be better to have this fix in a separate PR. Can we make an exception in this case? <3
